### PR TITLE
feat: Create CDD component [PAGE-006] ProblemList #257

### PR DESCRIPTION
**Pull request by HersonJose**
URL: 
https://ditmar.github.io/leetcode-spa/feature-leetcode-257/?path=/story/pages-problemlist--default

### CODE REVIEW

File: src/pages/ProblemList/ProblemList.constants.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+export const STATUS_ICON_MAP = {
+  solved: CheckCircleOutlineIcon,
+  attempted: RemoveCircleOutlineIcon,
+  unsolved: RadioButtonUncheckedIcon,
+} as const;
+
+export const STATUS_COLOR_MAP = {
+  solved: 'success',
+  attempted: 'warning',
```
Comment (Ditmar) [inline src/pages/ProblemList/ProblemList.constants.ts]: The `unsolved` status is mapped to 'neutral', which may not be consistent with the color scheme used in the rest of the application. Consider using a more descriptive color name or ensuring it aligns with the design.

**Suggestion:**
```typescript
Change 'neutral' to a more appropriate color name that reflects the intended design.
```

File: src/pages/ProblemList/ProblemList.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        <Typography
+          variant="h6"
+          component="h2"
+          sx={{
+            fontFamily: 'Inter, sans-serif',
+            fontWeight: 600,
+            fontSize: '16px',
+            color: '#1a1a1a',
+            letterSpacing: '-0.2px',
+          }}
```
Comment (Ditmar) [inline src/pages/ProblemList/ProblemList.tsx:144]: The inline color logic for the status icons could lead to duplication and make future changes harder. Consider using the `STATUS_COLOR_MAP` for consistency.

**Suggestion:**
```typescript
Replace the inline color logic with a reference to `STATUS_COLOR_MAP` to maintain consistency and reduce duplication.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
**Conclusion:** The implementation is mostly solid, but the handling of status colors needs to be addressed to ensure consistency and maintainability. Once these issues are resolved, the code can be approved.

File: src/pages/ProblemList/ProblemList.test.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import { describe, test, expect, vi } from 'vitest';
+
+import { ProblemList } from './ProblemList';
+
+vi.mock('./ProblemList.constants', async (importOriginal) => {
+  const actual = await importOriginal<typeof import('./ProblemList.constants')>();
+
+  const DummyIcon = ({ 'data-testid': testId }: { 'data-testid'?: string }) => (
+    <span data-testid={testId}>•</span>
+  );
```
Comment (Ditmar) [inline src/pages/ProblemList/ProblemList.test.tsx]: The mockProblems array should ensure that all properties match the expected types defined in the Problem interface. Using 'as const' for string literals is good, but ensure that all values are strictly typed.

**Suggestion:**
```typescript
Consider using TypeScript's type assertions to ensure that the mockProblems array strictly adheres to the Problem interface.
```

File: src/pages/ProblemList/ProblemList.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        >
+          Problems
```
Comment (Ditmar) [inline src/pages/ProblemList/ProblemList.tsx:146]: The iconColor assignment relies on string comparisons which could lead to issues if the status values are not strictly controlled. If a new status is added in the future, this logic may break.

**Suggestion:**
```typescript
Consider using a mapping object or a switch statement to handle icon colors based on status to improve maintainability and reduce the risk of errors.
```

File: src/pages/ProblemList/ProblemList.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                key={problem.id}
+                sx={itemButtonStyles(isSelected)}
+                onClick={() => onSelectProblem(problem.id)}
+              >
+                <Box sx={itemIconStyles}>
+                  <StatusIcon sx={{ color: iconColor, fontSize: '18px' }} />
+                </Box>
+
+                <Box sx={{ flexGrow: 1, overflow: 'hidden' }}>
+                  <Typography sx={titleTypographyStyles}>
```
Comment (Ditmar) [inline src/pages/ProblemList/ProblemList.tsx]: The component does not handle the case where the problems array is empty. While it may render without errors, it could lead to unexpected behavior if not properly managed.

**Suggestion:**
```typescript
Add a conditional rendering check to display a message or a placeholder when the problems array is empty.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
The code introduces a new ProblemList component with clear structure and organization. The use of constants for icons and colors enhances maintainability. However, there are some areas for improvement regarding type safety and potential edge cases in the component's behavior.
---

**Conclusion:** While the implementation is largely well-structured, the identified issues need to be addressed to ensure type safety and robustness in handling edge cases.

File: src/pages/ProblemList/ProblemList.test.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import { describe, it, expect, vi } from 'vitest';
+
+import { ProblemList } from './ProblemList';
+import { MOCK_PROBLEMS } from './ProblemList.constants';
+
+const defaultProps = {
+  problems: MOCK_PROBLEMS,
+  selectedProblemId: 1,
+  onSelectProblem: vi.fn(),
+};
```
Comment (Ditmar) [inline src/pages/ProblemList/ProblemList.test.tsx:16]: The mockProblems array should ensure that all properties match the expected types defined in the Problem interface. Using 'as const' for string literals is good, but ensure that all values are strictly typed.

**Suggestion:**
```typescript
Consider using TypeScript's type assertions to ensure that the mockProblems array strictly adheres to the Problem interface.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
Three comments remain unresolved, including critical issues regarding status color consistency and handling of empty problem arrays. Additionally, a new issue was identified regarding type safety in the mockProblems array.

### 🔄 Re-review: Previous Comments Status

**✅ Addressed (1):**
- `src/pages/ProblemList/ProblemList.tsx:66` — The implementation now includes a check for an empty problems array, displaying a message when no problems are found.

**❌ Still pending (4):**
- `src/pages/ProblemList/ProblemList.constants.ts:15` — The status 'unsolved' is still mapped to 'neutral' in the mock data, which does not align with the color scheme.
  > _Original comment:_ The `unsolved` status is mapped to 'neutral', which may not be consistent with the color scheme used in the rest of the application. Consider using a more descriptive color name or ensuring it aligns with the design.
- `src/pages/ProblemList/ProblemList.tsx:145` — The inline color logic for status icons remains unchanged, leading to potential duplication.
  > _Original comment:_ The inline color logic for the status icons could lead to duplication and make future changes harder. Consider using the `STATUS_COLOR_MAP` for consistency.
- `src/pages/ProblemList/ProblemList.test.tsx:12` — The mockProblems array does not strictly adhere to the Problem interface, which could lead to type safety issues.
  > _Original comment:_ The mockProblems array should ensure that all properties match the expected types defined in the Problem interface. Using 'as const' for string literals is good, but ensure that all values are strictly typed.
- `src/pages/ProblemList/ProblemList.tsx:147` — The iconColor assignment logic still relies on string comparisons, which poses a risk for future maintainability.
  > _Original comment:_ The iconColor assignment relies on string comparisons which could lead to issues if the status values are not strictly controlled. If a new status is added in the future, this logic may break.

---

**Conclusion:** The pull request cannot be approved due to unresolved critical issues regarding status color consistency and type safety in the mockProblems array.

**Review by Ditmar — APPROVED** [review]
All previous comments have been addressed. The handling of status colors has been improved for consistency, and the component now properly handles empty states. No new issues were found in the latest changes.

### 🔄 Re-review: Previous Comments Status

**✅ Addressed (5):**
- `src/pages/ProblemList/ProblemList.constants.ts:15` — The status color mapping has been updated to use 'disabled' instead of 'neutral', aligning it with the design.
- `src/pages/ProblemList/ProblemList.tsx:144` — The inline color logic has been replaced with a reference to `STATUS_COLOR_MAP`, improving maintainability.
- `src/pages/ProblemList/ProblemList.test.tsx:12` — The mockProblems array has been updated to ensure all properties match the expected types.
- `src/pages/ProblemList/ProblemList.tsx:146` — The logic for assigning icon colors has been refactored to use a mapping object, improving maintainability.
- `src/pages/ProblemList/ProblemList.tsx:66` — A conditional rendering check has been added to display a message when the problems array is empty.

---

**Conclusion:** All critical and major issues from previous reviews have been resolved, and no new issues were identified.
